geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/AttributeIndex.scala

```diff
@@ -158,75 +109,165 @@ trait AttributeIndex[DS <: GeoMesaDataStore[DS, F, W], F <: WrappedFeature, W, R
+    override def supports(sft: SimpleFeatureType): Boolean = sft.getDtgField.isDefined
 
-  // store the first 12 hex chars of the time - that is roughly down to the minute interval
-  def timeToBytes(t: Long) = typeRegistry.encode(t).substring(0, 12).getBytes(StandardCharsets.UTF_8)
+    override val indexKeyLength: Int = 12
 
-  // rounds up the time to ensure our range covers all possible times given our time resolution
-  private def roundUpTime(time: Array[Byte]): Array[Byte] = {
-    // find the last byte in the array that is not 0xff
-    var changeIndex: Int = time.length - 1
-    while (changeIndex > -1 && time(changeIndex) == 0xff.toByte) { changeIndex -= 1 }
+    override def toIndexKey(sft: SimpleFeatureType): (SimpleFeature) => Array[Byte] = {
+      val dtgIndex = sft.getDtgIndex.getOrElse(-1)
+      (feature) => {
+        val dtg = feature.getAttribute(dtgIndex).asInstanceOf[Date]
+        timeToBytes(if (dtg == null) { 0L } else { dtg.getTime })
+      }
+    }
 
-    if (changeIndex < 0) {
-      // the array is all 1s - it's already at time max given our resolution
-      time
-    } else {
-      // increment the selected byte
-      time.updated(changeIndex, (time(changeIndex) + 1).asInstanceOf[Byte])
+    override def getRanges(sft: SimpleFeatureType,
+                           filter: Filter,
+                           explain: Explainer): Iterator[(Array[Byte], Array[Byte])] = {
+      val intervals = sft.getDtgField.map(FilterHelper.extractIntervals(filter, _)).getOrElse(FilterValues.empty)
+      intervals.values.iterator.map { case (lo, hi) =>
+        (timeToBytes(lo.getMillis), roundUpTime(timeToBytes(hi.getMillis)))
+      }
+    }
+
+    // store the first 12 hex chars of the time - that is roughly down to the minute interval
+    private def timeToBytes(t: Long): Array[Byte] =
+      typeRegistry.encode(t).substring(0, 12).getBytes(StandardCharsets.UTF_8)
+
+    // rounds up the time to ensure our range covers all possible times given our time resolution
+    private def roundUpTime(time: Array[Byte]): Array[Byte] = {
+      // find the last byte in the array that is not 0xff
+      var changeIndex: Int = time.length - 1
+      while (changeIndex > -1 && time(changeIndex) == 0xff.toByte) { changeIndex -= 1 }
+
+      if (changeIndex < 0) {
+        // the array is all 1s - it's already at time max given our resolution
+        time
+      } else {
+        // increment the selected byte
+        time.updated(changeIndex, (time(changeIndex) + 1).asInstanceOf[Byte])
+      }
     }
   }
+}
+
+object AttributeIndex {
+
+  val NullByte: Byte = 0
+  val NullByteArray  = Array(NullByte)
+
+  val typeRegistry   = LexiTypeEncoders.LEXI_TYPES
+
+  // store 2 bytes for the index of the attribute in the sft - this allows up to 32k attributes in the sft.
+  def indexToBytes(i: Int): Array[Byte] = Array((i << 8).asInstanceOf[Byte], i.asInstanceOf[Byte])
```

Comment:
If there are more than 32k attributes, does this 'bin' things by the first two bytes?